Refactor /parse, remove validation, support batch

hkeeler opened the pull request:
This PR does a few things:
1. Removes `/standardize` resource. (Closes #28)
2. Refactor's `/parse` resource.
   1. Response JSON has been refactored to an array of address parts, rather than explicit key/values. (Closes #23)
   2. Supports `POST` of up to 5000 address for batch parsing. (Closes #30)
   3. Removes validation code that was causing heartache within the geocoder.
      - **Note:** Some form of validation will likely return in the future...
3. Adds more tests.

hkeeler committed:
Refactor to single simplified /parse resource

- Removes /standardize resource
- Removes "validate" param and validation functionality
- Returns array of address parts instead of key/val pair
  based on usaddress address part type
- Returns parts in order they appear in full string, not
  alphabetic order of dict keys

hkeeler committed:
Adds batch parsing.  Still needs tests.

hkeeler committed:
Adds first set of batch parsing tests

hkeeler committed:
Splits batch response into "parsed" and "failed"

hkeeler committed:
400s when 'addresses' empty, null, or too full

hkeeler committed:
Adds test for 404s

hkeeler commented:
Looks like one of the tests is failing in Travis:
- https://travis-ci.org/cfpb/grasshopper-parser/jobs/90343252#L156

_Works on my machine_.  I'll look into it.

hkeeler commented:
We'll this is _interesting_.  One of the new tests checks that the `failed` array is populated when an unparsable address string is passed in.  This works fine when run using `nosetests`, but when run via `tox` (which itself calls `nose`), the test **fails** because the address parsing **works**.

I'm using the bad address we saw during the demo last week:

```
5 Arapahoe Plaza El Paso TX 88530
```

...which parses to:

``` python
{u'input': u'5 Arapahoe Plaza El Paso TX 88530',
  u'parts': [
    {u'type': u'BuildingName', u'value': u'5 Arapahoe Plaza El'},
    {u'type': u'PlaceName', u'value': u'Paso'},
    {u'type': u'StateName', u'value': u'TX'},
    {u'type': u'ZipCode', u'value': u'88530'}]}
```

It clearly doesn't parse _well_, but it does parse.  So, the question is, _why_?  I'm guessing I have newer version of `usaddress` in my local virtualenv, probably from the unmerged https://github.com/cfpb/grasshopper-parser/pull/26.  I'm going to merge that PR into this one and see if it fixes it.

hkeeler committed:
Merge branch 'master' of github.com:cfpb/grasshopper-parser into merge-standardize

hkeeler commented:
@awolfe76, thanks for the merge of #26.  It seems to have fixed both the failing unit test _and_ the Coveralls integration!

awolfe76 commented:
That's great! Glad that worked.

hkeeler committed:
Adds rules.yaml for mapping usaddress address parts

hkeeler committed:
Fixes address part id variables in rules.yaml

hkeeler committed:
First pass at concat-ing address number and street name address parts

hkeeler committed:
Adds "profile" support to batch parsing

hkeeler committed:
Adds validation of "required" address parts

hkeeler committed:
Move parser logic into its own class

hkeeler committed:
Removes "method" param since its now set on parser constructor

hkeeler committed:
Expands test suite; Adds assert_equals hack for diffing log lists

hkeeler commented:
This PR's sister PR https://github.com/cfpb/grasshopper/pull/195 is now ready for review.  They'll need to be merged at the same time, otherwise we'll have a broken geocoder.

I do still need to add docs for how the API has changed.  I will take care of that as soon as I'm back on Monday.

jmarin commented on tests/tests.py:
```diff
@@ -55,194 +159,203 @@ def test_internal_error(self):
+        resp_data = json.loads(resp.data)
+
+        assert_equals(200, resp.status_code)
+        assert_equals(resp_data['failed'][0], '5 Arapahoe Plaza El Paso TX 88530')
```
@hkeeler This assertion fails when running the tests on this branch. 

```
======================================================================
ERROR: API: POST /parse -> 200 with good and bad address strings
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Python/2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "~/Development/grasshopper-parser/tests/tests.py", line 322, in test_parse_batch_with_failed_parse
    assert_equals(resp_data['failed'][0], '5 Arapahoe Plaza El Paso TX 88530')
IndexError: list index out of range

----------------------------------------------------------------------
```

hkeeler commented:
@jmarin, the failing test sounds very similar to the issues described [above](https://github.com/cfpb/grasshopper-parser/pull/33#issuecomment-155511836).  You probably just need to get the latest Python packages introduced on #26.  If you haven't done so already, `pip install` and try it again:

```
pip install -r requirements.txt
pip install -r tests/requirements.txt
```

...and also from #26, you may have watch out for:

> **Warning:** The upgrade from Coverage 3.x to 4.x breaks compatibility with the `.coverage` file format.  If you have a `.coverage` file in your project root, you'll need to delete it before running `coverage` or `tox`.

hkeeler committed:
Fixes all flake8 recommendations...mostly whitespace

hkeeler committed:
Converts address part type name from "type" to "code"

hkeeler commented:
I've converted the "type" field to "code" to get around the `type` keyword ugliness on the Scala side.  I also fixed all of the `flake8` recommendations.  This should be good to go.
